sdk/eventhub/azure-eventhubs/azure/eventhub/_client_base.py

```diff
@@ -0,0 +1,376 @@
+        """
+        Get partition ids of the specified EventHub.
+
+        :rtype: list[str]
+        :raises: :class:`EventHubError<azure.eventhub.EventHubError>`
+        """
+        return self.get_properties()['partition_ids']
+
+    def get_partition_properties(self, partition):
+        # type:(str) -> Dict[str, Any]
+        """
+        Get properties of the specified partition async.
+        Keys in the details dictionary include:
+
+            - event_hub_path
+            - id
+            - beginning_sequence_number
+            - last_enqueued_sequence_number
+            - last_enqueued_offset
+            - last_enqueued_time_utc
+            - is_empty
+
+        :param partition: The target partition id.
+        :type partition: str
+        :rtype: dict
+        :raises: :class:`EventHubError<azure.eventhub.EventHubError>`
+        """
+        mgmt_msg = Message(application_properties={'name': self.eh_name,
+                                                   'partition': partition})
+        response = self._management_request(mgmt_msg, op_type=MGMT_PARTITION_OPERATION)
+        partition_info = response.get_data()
+        output = {}
+        if partition_info:
+            output['event_hub_path'] = partition_info[b'name'].decode('utf-8')
+            output['id'] = partition_info[b'partition'].decode('utf-8')
+            output['beginning_sequence_number'] = partition_info[b'begin_sequence_number']
+            output['last_enqueued_sequence_number'] = partition_info[b'last_enqueued_sequence_number']
+            output['last_enqueued_offset'] = partition_info[b'last_enqueued_offset'].decode('utf-8')
+            output['last_enqueued_time_utc'] = datetime.datetime.utcfromtimestamp(
+                float(partition_info[b'last_enqueued_time_utc'] / 1000))
+            output['is_empty'] = partition_info[b'is_partition_empty']
+        return output
+
+    def close(self):
+        # type:() -> None
+        self._conn_manager.close_connection()
+
+
+class ConsumerProducerMixin(object):
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    def _check_closed(self):
+        if self.closed:
+            raise EventHubError("{} has been closed. Please create a new one to handle event data.".format(self._name))
+
+    def _open(self):
+        """Open the EventHubConsumer/EventHubProducer using the supplied connection.
+
+        """
+        # pylint: disable=protected-access
+        if not self.running:
+            if self._handler:
+                self._handler.close()
+            self._create_handler()
+            self._handler.open(connection=self._client._conn_manager.get_connection(  # pylint: disable=protected-access
+                self._client._address.hostname,
+                self._client._create_auth()
+            ))
+            while not self._handler.client_ready():
+                time.sleep(0.05)
```

Comment:
Is the mixin called from both sync and async code? If the latter, never `time.sleep()` inside the eventloop.

Reply:
No this is sync only

Reply:
And can this sleep forever? There is no obvious timeout plumbed through here...